### PR TITLE
separate out mangleType()

### PR DESCRIPTION
**Pull request by WalterBright**
The Mangler code is very tangled up. This is an attempt at untangling it somewhat, by pulling all the Type mangling into mangleType(). I unsuccessfully tried to minimize the diffs, this was as small as I could get.

### CODE REVIEW

**Review by thewilsonator — CHANGES_REQUESTED** [review]
Split the `mangleParameter` changes into a separate PR, that will cut ~200 lines for the diff

**Review by maxhaton — APPROVED** [review]

**Comment by WalterBright** [comment]
@thewilsonator it's a good idea, but since it's already passing it's better to move on.

**Comment by ibuclaw** [comment]
This caused a build regression.
```d
0x22591ad onAssertError
	../../../../libphobos/libdruntime/core/exception.d:558
0xa8158d dmd.dmangle.mangleType(Type, ubyte, OutBuffer*, ref Backref).visitWithMask(Type, ubyte)
	../../gcc/d/dmd/dmangle.d:417
0xa8158d dmd.dmangle.mangleType(Type, ubyte, OutBuffer*, ref Backref).visitWithMask(Type, ubyte)
	../../gcc/d/dmd/dmangle.d:254
0xa815ac dmd.dmangle.mangleType(Type, ubyte, OutBuffer*, ref Backref)
	../../gcc/d/dmd/dmangle.d:422
0xa810b3 dmd.dmangle.mangleFuncType(TypeFunction, TypeFunction, ubyte, Type, OutBuffer*, ref Backref)
	../../gcc/d/dmd/dmangle.d:512
0xa815ac dmd.dmangle.mangleType(Type, ubyte, OutBuffer*, ref Backref)
	../../gcc/d/dmd/dmangle.d:422
0xa815fb mangleToBuffer(Type*, OutBuffer*)
	../../gcc/d/dmd/dmangle.d:43
```

File: compiler/src/dmd/dmangle.d [hunk 1]
```diff
@@ -231,235 +241,376 @@ unittest
...
+        case Tcomplex64:
+        case Tcomplex80:
+
+        case Tbool:
+        case Tchar:
+        case Twchar:
+        case Tdchar:
+
+        case Tint128:
+        case Tuns128:    visitType(t); break;
```
Comment (ibuclaw) [inline compiler/src/dmd/dmangle.d:393]: No.

File: compiler/src/dmd/dmangle.d [hunk 2]
```diff
@@ -231,235 +241,376 @@ unittest
...
+        case Ttuple:     visitTypeTuple     (t.isTypeTuple());      break;
+        case Tnull:      visitTypeNull      (t.isTypeNull());       break;
+        case Tvector:    visitTypeVector    (t.isTypeVector());     break;
+        case Tnoreturn:  visitTypeNoreturn  (t.isTypeNoreturn);     break;
+
+        case Terror:
+            break;      // ignore errors
+
+        default:
+            assert(0);
```
Comment (ibuclaw) [inline compiler/src/dmd/dmangle.d:417]: No! This is not equivalent to the original.

File: compiler/src/dmd/dmangle.d [hunk 3]
```diff
@@ -231,235 +241,376 @@ unittest
...
-        mangleType(t);
+        scope Mangler v = new Mangler(buf, &backref);
+        v.mangleSymbol(s);
     }
 
-    override void visit(Type t)
+    void visitType(Type t)
     {
         tyToDecoBuffer(buf, t.ty);
     }
```
Comment (ibuclaw) [inline compiler/src/dmd/dmangle.d:265]: The original handled _all_ other types permissively.

**Pull request merged**